x/pool/bucketized_bool_array_pool.gen.go

```diff
@@ -0,0 +1,158 @@
+
+package pool
+
+import (
+	"fmt"
+
+	"sort"
+
+	"github.com/uber-go/tally"
+)
+
+// BoolArrayBucket specifies a bucket.
+type BoolArrayBucket struct {
+	// Capacity is the size of each element in the bucket.
+	Capacity int
+
+	// Count is the number of fixed elements in the bucket.
+	Count int
+
+	// Options is an optional override to specify options to use for a bucket,
+	// specify nil to use the options specified to the bucketized pool
+	// constructor for this bucket.
+	Options *BoolArrayPoolOptions
+}
+
+// boolArrayBucketByCapacity is a sortable collection of pool buckets.
+type boolArrayBucketByCapacity []BoolArrayBucket
+
+func (x boolArrayBucketByCapacity) Len() int {
+	return len(x)
+}
+
+func (x boolArrayBucketByCapacity) Swap(i, j int) {
+	x[i], x[j] = x[j], x[i]
+}
+
+func (x boolArrayBucketByCapacity) Less(i, j int) bool {
+	return x[i].Capacity < x[j].Capacity
+}
+
+type boolArrayBucketPool struct {
+	capacity int
+	pool     *BoolArrayPool
+}
+
+// BucketizedBoolArrayPool is a bucketized value pool.
+type BucketizedBoolArrayPool struct {
+	sizesAsc          []BoolArrayBucket
+	buckets           []boolArrayBucketPool
+	maxBucketCapacity int
+	opts              *BoolArrayPoolOptions
+	alloc             func(capacity int) []bool
+	maxAlloc          tally.Counter
+}
+
+// NewBucketizedBoolArrayPool creates a bucketized object pool.
+func NewBucketizedBoolArrayPool(sizes []BoolArrayBucket, opts *BoolArrayPoolOptions) *BucketizedBoolArrayPool {
+	if opts == nil {
+		opts = NewBoolArrayPoolOptions()
+	}
+
+	sizesAsc := make([]BoolArrayBucket, len(sizes))
+	copy(sizesAsc, sizes)
+	sort.Sort(boolArrayBucketByCapacity(sizesAsc))
+
+	var maxBucketCapacity int
+	if len(sizesAsc) != 0 {
+		maxBucketCapacity = sizesAsc[len(sizesAsc)-1].Capacity
+	}
+
+	return &BucketizedBoolArrayPool{
+		opts:              opts,
+		sizesAsc:          sizesAsc,
+		maxBucketCapacity: maxBucketCapacity,
+		maxAlloc:          opts.MetricsScope().Counter("alloc-max"),
+	}
+}
+
+// Init initializes the bucketized pool.
+func (p *BucketizedBoolArrayPool) Init(alloc func(capacity int) []bool) {
+	buckets := make([]boolArrayBucketPool, len(p.sizesAsc))
+	for i := range p.sizesAsc {
+		size := p.sizesAsc[i].Count
+		capacity := p.sizesAsc[i].Capacity
+
+		opts := p.opts
+		if perBucketOpts := p.sizesAsc[i].Options; perBucketOpts != nil {
+			opts = perBucketOpts
+		}
+
+		opts = opts.SetSize(size)
+		scope := opts.MetricsScope()
+		if scope != nil {
+			opts = opts.SetMetricsScope(scope.Tagged(map[string]string{
+				"bucket-capacity": fmt.Sprintf("%d", capacity),
+			}))
+		}
+
+		buckets[i].capacity = capacity
+		buckets[i].pool = NewBoolArrayPool(opts)
+		buckets[i].pool.Init(func() []bool {
+			return alloc(capacity)
+		})
+	}
+	p.buckets = buckets
+	p.alloc = alloc
+}
+
+// Get gets a value from the pool.
+func (p *BucketizedBoolArrayPool) Get(capacity int) []bool {
+	if capacity > p.maxBucketCapacity {
+		p.maxAlloc.Inc(1)
+		return p.alloc(capacity)
```

Comment:
Should we allocate the requested capacity or should we have well defined bucket sizes and go up one size?

e.g.
```
1 << 2
1 << 3
1 << 4
1 << 5
1 << 6
etc.
```

I was thinking about the case where you get multiple requests to allocate increasingly larger slices but only by a small margin resulting in many allocs. Might be too much of an edge case to handle though.

Reply:
That is indeed the intention, i.e., having bucket sizes that are exponentially increasing by a factor of 2. However this should be configured externally as there may be usecases where this is not desirable. For our usecases yes, a factor of 2 should work just fine.